### ISSUE TITLE
Idea: Rewrite library to use TypeScript ?

### ISSUE THREAD

**Issue by csowada** [openhab/openhab-js#41]
I'm a fan of Visual Studio Code, Intellisense and Typescript. So the auto complete feature i not working correct with the current version. 
In your eyes, does it make sense to switch the codebase to TypeScript?
I have already done this for openHAB 3.1 and would like to go back to the standard. But for the standard module I would want less features but good documentation and support for the common platforms.

I would like to help, but only if this could go back into the standard.

**Comment by Confectrician** [comment]
Hey @csowada 

This is already in discussion with a different approach of solution.

See: https://github.com/openhab/openhab-webui/pull/1237#issuecomment-997090501

**Comment by csowada** [comment]
Thank you @Confectrician for the hint. It goes into the same direction to support Visual Studio Code in the best way. TypeScript is only one way to reach it.

**Comment by digitaldan** [comment]
So there are two options, one is to rewrite the whole library in TS, the other is to provide Type Declarations file(s) to just provide the API interfaces of the library ( as was shown in https://github.com/openhab/openhab-webui/pull/1237#issuecomment-997090501)

This second option is an officially support one by TS, see:
https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html
Whats nice is we can automate this from the JSDoc comments we are already using

And also day.js, which is a popular time library that does this
https://github.com/iamkun/dayjs/tree/dev/types

**Comment by Confectrician** [comment]
> Whats nice is we can automate this from the JSDoc comments we are already using

We can automate it and it is up to date by default, when releasing a new version.
For me there is no reason doing it the manual way then. 🙂

**Comment by csowada** [comment]
I was working on it today, but we need to adjust some js-doc comments and split files to resolve all issues. The namespaces will be ignored by TypeScript converter, but we use modules. So I see no benefit.I prefer to load the modules on my own instead of global preloaded.

**Comment by digitaldan** [comment]
> I was working on it today, but we need to adjust some js-doc comments and split files to resolve all issues. 

Thanks, thats great.  To be honest, i rushed through the JSDocs without really understanding it to get something out the door for 3.2 , but it's high on my list to improve so our docs are more readable.  I assume there would also be some changes to our namespace and class structure.

**Comment by csowada** [comment]
Here is my current status, it works with the type definition. But this is just my working branch. It contains some rubbish and test.  One problem is that the generator not create correct index.d.ts files or I'm not aware how to handle it correct. This is the reason for the ``dist-overwrite`` folder.

https://github.com/csowada/openhab-js

To start a new project just do ...

```shell
npm init
npm install github:csowada/openhab-js#d_ts_optimize
```

afterward create a index.js

```typescript
const { log, rules, triggers } = require("openhab");
const logger = log("mylogger");

rules.JSRule({
    name: "",
    description: "ohh",
    triggers: [
        triggers.ItemStateChangeTrigger("ItemName")
    ],
    execute: () => {
        logger.info("Hello World");
    }
});
```

**Comment by csowada** [comment]
I think a TypeScript port is still a good thing. So nearly finished a TypeScript version. I found maybe so inconsistencies in the scripts. It is easier to identify null errors and wrong references as it is more strict. But the work is not finished yet, but I want to inform other that I'm working on it.

https://github.com/csowada/openhab-js/tree/ts-fork

**Comment by digitaldan** [comment]
Thanks @csowada , from a language framework point of view , i'm impartial to plain ES6 vs typscript, but i wonder what others think here?  My main interests are that we have a framework that is easy to maintain, remains simple a straight forward,  and is written so others can contribute to it as easily as possible.    @jpg0  , i would very much like your thoughts on this.

**Comment by csowada** [comment]
I understand that plain ES6 would be nice, but I see the advantage by a more strict framework. The code will be transpiled by tsc and can be adjusted to the GraalVMs requirements. You can still use JS for your own rules etc. And the TS overhead is nearly nothing. And we can also describe global variables and later on Java classes. I've spend much time on my previous library to support many openhab Java classes.

**Comment by jpg0** [comment]
My view on this is:

- I do see technical value in Typescript; almost all teams that I have been involved with professionally that started out with JS ended up using TS. I myself ultimately leave JS when the system gets too complex as I find that I need more help from the compiler. On the flip side; I would expect most users not to be dealing with this level of complexity in scripts to need this (but there will be some; I used to be one).
- I personally don't have enough experience with TS to know how compatible (or maybe, optional) it is with standard JS. For example:

  * Can users just code in pure JS and there be no impact with the library being is TS? (I believe that the answer is yes)
  * Can library developers commit changes, or new parts of the library which are in pure JS? (I believe that the answer is no, or technically yes, but practically no)
  * Can we make the switch in either direction, and retain compatibility? (I believe that the answer is that we can go JS->TS without breaking things for users, but not the other way around)

I guess that ultimately this comes down to whether we have expertise to maintain a library here in TS rather than JS. Right now we have @digitaldan doing a great job here, and me offering a little advice but that's about all(!). Therefore my view is that this decision should be made by whoever is actually committing code to the project, and whether they themselves want to be writing it in JS or TS.

**Comment by csowada** [comment]
One important thing is, that normal users don't need TS, they can use JS. But VSCode can help with correct Auto Complete etc because of the advanced information.
I'll answer your questions later on, we are still in Christmas Mode :-)

**Comment by csowada** [comment]
> My view on this is:
> 
> * I do see technical value in Typescript; almost all teams that I have been involved with professionally that started out with JS ended up using TS. I myself ultimately leave JS when the system gets too complex as I find that I need more help from the compiler. On the flip side; I would expect most users not to be dealing with this level of complexity in scripts to need this (but there will be some; I used to be one).
> * I personally don't have enough experience with TS to know how compatible (or maybe, optional) it is with standard JS. For example:
>   
>   * Can users just code in pure JS and there be no impact with the library being is TS? (I believe that the answer is yes)

Yes, the user can choose between JavaScript and TypeScript. When the library is built, TypeScript is translated into JavaScript. Thus, JavaScript is always used in GraalVM.

>   * Can library developers commit changes, or new parts of the library which are in pure JS? (I believe that the answer is no, or technically yes, but practically no)

Yes, developers could also use pure JavaScript. It would also be translated with. But it would not be very useful for a library. But every JavaScript developer should understand and be able to use the used types.

Here is an example of the JavaScript/TypeScript difference. In most cases only the type is specified in the parameters and the return value including null values.

```javascript
export const log = (_name) => {
    return new Logger(_name);
}
```
or 
```typescript
export const log = (_name: string): Logger => {
    return new Logger(_name);
}
```

>   * Can we make the switch in either direction, and retain compatibility? (I believe that the answer is that we can go JS->TS without breaking things for users, but not the other way around)
> 

> I guess that ultimately this comes down to whether we have expertise to maintain a library here in TS rather than JS. Right now we have @digitaldan doing a great job here, and me offering a little advice but that's about all(!). Therefore my view is that this decision should be made by whoever is actually committing code to the project, and whether they themselves want to be writing it in JS or TS.

I think it's important to know that TypeScript is actually JavaScript plus types. And I see the types as a part of documentation and quality management. The strict types help to identify errors in the library faster. And also the JavaScript end user gets these hints in Visual Studio Code.
In addition, we can also still describe important Java classes and their methods in TypeScript.
Additionally, with TypeScript you can already rely on modules like in ES6-ES2021 and have this compiled to CommonJS require(). Thus, the library is also already up to date.

It is best to compare the generated JavaScript files and the TypeScript files to get a feeling for the difference.

Example:

https://github.com/csowada/openhab-js/blob/ts-fork/dist/cache.js

https://github.com/csowada/openhab-js/blob/ts-fork/src/cache.ts

**Comment by jpg0** [comment]
@csowada I think that verifies all my assumptions. In that case my assessment still stands - whilst there are clear technical merits, it's up to the primary person(s) maintaining the code.

**Comment by csowada** [comment]
For the moment I'll stay at my own library https://github.com/csowada/openhab-jsscripting-base to bring my productive system to v3.2 . So I can focus on the other changes in 3.2 release.
I'll continue my work on a TS fork if we have a decision that TS is the way to go.

**Comment by digitaldan** [comment]
thanks for your thoughts @jpg0 .  @csowada, thanks as well.  I'm on the fence here. The thing i'm contemplating is that we are talking about converting the whole project to TS just for a single feature, intellisense support.  I understand all the other benefits of TS, i have worked on many teams that use TS either b/c  A) they are all primarily Java/Scala developers or  B) were using angular.    But with ES6 and classes , some of the benefits of TS have lessened and I am seeing a return to standard javascript.

I still want to think it over some before making a decision on this as once we make it, it will be hard to go back from it.  The other think i want to confirm before making a decision is if we can inject the imports through the VSCode OH extension, as most users will not be using `const { item, things } = require('openhab);` at the top of their files as we inject this automatically.

**Comment by csowada** [comment]
Hello @digitaldan,

I understand your

**Comment by csowada** [comment]
Hi @digitaldan, I completely understand your concerns. All library developers have to accept this decision. But for example in my translations to TypeScript I also came across inconsistencies. There are two items, the wrapper item and the Java Native item. I think this was mixed up in places in the code.

But I don't see TypeScript and ES6/2021 in competition. It makes it much easier to implement cool new language features in ES6 without hacks and polyfills. So a win-win situation for TS and ES. And with d.ts files we can also describe the important Java files of openHAB. With the changeover, one would then also switch from CommonJS to ES modules within the library. So instead of require() then import. This would bring us completely to ES6-2021.

As I already mentioned, I already have my own TypeScript version that does not wrap Java items, but only provides the most important helper functions. And of course a lot of type definitions. But most of the functions were taken from this library and rewritten in TS. But I also don't want to force my opinion on anyone. Maybe it should run in parallel so that we can decide later without pressure.

But for my own rules (only!) in TS I'm thinking about switching back to JS, because the well described TS library is enough TS to work with Intellisense. I change my rules much more often than the library. In that case the transpile step has no benefit for my rules.

>I still want to think it over some before making a decision on this as once we make it, it will be hard to go back from it. The other think i want to confirm before making a decision is if we can inject the imports through the VSCode OH extension, as most users will not be using const { item, things } = require('openhab); at the top of their files as we inject this automatically.

Here it should be sufficient to provide the namespaces directly in a global.d.ts. This way they can be referenced directly when using the library. But a parallel load with require() should also not cause any problems. Here is my globald.ts as an example.

https://github.com/csowada/openhab-jsscripting-base/blob/main/%40types/openhab-3.0.0.d.ts

**Comment by rkoshak** [comment]
>  but that's about all(!)

I hope to become a contributor but I don't know how much I'll be able to take on. 

Lots to learn first before I start pushing code though.

Anyway, the only thing I have to add to this is to think about who is most likely to contribute to this library. It's going to be the more technical openHAB users who do it. If switching to TypeScript puts up barriers to people whose only exposure to this stuff is writing rules in openHAB I fear that might cut out a good number of potential contributors off. That's something to consider, especially since we've been burned by this once before already (PaperUI went away because the contributor who wrote it chose a framework that no one else in OH knew and could support). 

I don't know enough about either yet to know if that's a real concern here or not. But I *am* one of those users whose only real exposure to JS coding is through openHAB rules, UI rules at that.

**Comment by csowada** [comment]
It's important to know that TypeScript is just JavaScript + stronger Types. I'm sure that it is easy for interested people to apply changes to the library. If the whole library is transfered to TS it is more copy and paste. If you add a new function, it is just important to think about input parameters and what you will return. Specially optional and null return values are cleaner defined like @Null annotations in Java .

We should not force normal library end users to use TypeScript. They can write all there rules in JavaScript.

**Comment by florian-h05** [comment]
Closing as there has not been any activity for over six months and type definitions have been added to the library.
